xtra-libp2p/src/dialer.rs

```diff
@@ -1,107 +1,124 @@
-use crate::ConnectionStats;
 use crate::Endpoint;
-use crate::GetConnectionStats;
 use anyhow::anyhow;
 use async_trait::async_trait;
 use libp2p_core::Multiaddr;
+use libp2p_core::PeerId;
 use std::time::Duration;
-use tokio_tasks::Tasks;
 use xtra::Address;
 use xtra_productivity::xtra_productivity;
-use xtras::SendInterval;
 
-/// How often we check whether we are still connected
-pub const CHECK_INTERVAL: Duration = Duration::from_secs(5);
+/// If we're not connected by this time, try again.
+pub const CONNECTION_TIMEOUT: Duration = Duration::from_secs(5);
 
 /// xtra actor that takes care of dialing (connecting) to an Endpoint.
 ///
 /// Periodically polls Endpoint to check whether connection is still active.
 /// Should be used in conjunction with supervisor maintaining resilient connection.
 pub struct Actor {
-    tasks: Tasks,
     endpoint: Address<Endpoint>,
     connect_address: Multiaddr,
+    connected: bool,
+    peer_id: Option<PeerId>,
     stop_reason: Option<Error>,
 }
 
 impl Actor {
     pub fn new(endpoint: Address<Endpoint>, connect_address: Multiaddr) -> Self {
         Self {
-            tasks: Tasks::default(),
             endpoint,
             connect_address,
+            connected: false,
+            peer_id: None,
             stop_reason: None,
         }
     }
 
-    /// Returns error if we cannot access the Endpoint or if the peer is not
-    /// connected to the Endpoint.
-    async fn check_connection_active_in_endpoint(&self) -> Result<(), Error> {
-        let ConnectionStats {
-            connected_peers, ..
-        } = self
-            .endpoint
-            .send(GetConnectionStats)
-            .await
-            .map_err(|_| Error::NoEndpoint)?;
-
-        connected_peers
-            .contains(
-                &self
-                    .connect_address
-                    .clone()
-                    .extract_peer_id()
-                    .ok_or(Error::InvalidPeerId)?,
-            )
-            .then(|| ())
-            .ok_or(Error::ConnectionDropped)
-    }
-
     async fn connect(&self) -> Result<(), Error> {
         self.endpoint
             .send(Connect(self.connect_address.clone()))
             .await
             .map_err(|_| Error::NoEndpoint)?
             .map_err(|e| Error::Failed { source: anyhow!(e) })
     }
+
+    fn stop_with_error(&mut self, e: Error, ctx: &mut xtra::Context<Self>) {
+        self.stop_reason = Some(e);
+        ctx.stop();
+    }
 }
 
 #[async_trait]
 impl xtra::Actor for Actor {
     type Stop = Error;
 
     async fn started(&mut self, ctx: &mut xtra::Context<Self>) {
-        let this = ctx.address().expect("we are alive");
+        match self
+            .connect_address
+            .clone()
+            .extract_peer_id()
+            .ok_or(Error::InvalidPeerId)
+        {
+            Ok(peer_id) => self.peer_id = Some(peer_id),
+            Err(e) => {
+                self.stop_with_error(e, ctx);
+            }
+        }
```

Comment:
It's impossible to make `new` fallible instead with the current supervisor. I'd be very open to support such construct, as it would reduce some awkwardness in these constructors, but until we support that there's not much I can do 😭

Reply:
Good point. Should be easy to support it on the supervisor. I'll have a go.